### PR TITLE
Add TestPropertiesDrivenEncryption

### PR DESCRIPTION
**Pull request by andersonm-1**

### CODE REVIEW

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+ *  - ENCRYPT_COLUMNS_AND_FOOTER_CTR:   Encrypt two columns and the footer, with different
+ *                                  keys. Use the alternative (AES_GCM_CTR_V1) algorithm.
+ *  - WRITE_PLAINTEXT:   Do not encrypt anything
+ *
+ *
+ *
+ * The read sample uses each of the following decryption configurations to read every
+ * encrypted files in the input directory:
+ *
+ *  - DECRYPT_COLUMNS_AND_FOOTER:   Decrypt using key retriever that holds the keys of
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: lets call it DECRYPT_WITH_KEY_RETRIEVER

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+ *                                  two encrypted columns and the footer key.
+ *  - READ_PLAINTEXT:   Do not decrypt anything.
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: lets call it NO_DECRYPTION

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(TestPropertiesDrivenEncryption.class);
+
+  @Rule
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  @Rule
+  public ErrorCollector errorCollector = new ErrorCollector();
+
+  private static final Base64.Encoder encoder = Base64.getEncoder();
+  private static final byte[] FOOTER_ENCRYPTION_KEY = new String("0123456789012345").getBytes();
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: lets call these "master keys", FOOTER_MASTER_KEY, etc, to diff this from DEKs in previous tests

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+  private void addErrorToErrorCollectorAndLog(String errorMessage, EncryptionConfiguration encryptionConfiguration,
+                                                     DecryptionConfiguration decryptionConfiguration) {
+    String fullErrorMessage = String.format("%s - %s Error: %s",
+      encryptionConfiguration, decryptionConfiguration, errorMessage);
+
+    errorCollector.addError(new Throwable(fullErrorMessage));
+    LOG.error(fullErrorMessage);
+  }
+
+  public enum EncryptionConfiguration {
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: please move these enums to the beginning of the class, makes it easier to understand the rest

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+ * encryption configurations:
+ *
+ *  - ENCRYPT_COLUMNS_AND_FOOTER:   Encrypt two columns and the footer, with different
+ *                                  keys.
+ *  - ENCRYPT_COLUMNS_PLAINTEXT_FOOTER:   Encrypt two columns, with different keys.
+ *                                  Do not encrypt footer (to enable legacy readers)
+ *                                  - plaintext footer mode.
+ *  - ENCRYPT_COLUMNS_AND_FOOTER_CTR:   Encrypt two columns and the footer, with different
+ *                                  keys. Use the alternative (AES_GCM_CTR_V1) algorithm.
+ *  - WRITE_PLAINTEXT:   Do not encrypt anything
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: lets call it NO_ENCRYPTION

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+        + "required boolean boolean_field; "
+        + "required int32 int32_field; "
+        + "required float float_field; "
+        + "required double double_field; "
+        + "} ");
+
+    SimpleGroupFactory f = new SimpleGroupFactory(schema);
+
+    for (Map.Entry<EncryptionConfiguration, Configuration> encryptionConfigurationEntry : encryptionPropertiesMap.entrySet()) {
+      KeyToolkit.removeCacheEntriesForToken(KeyToolkit.DEFAULT_ACCESS_TOKEN);
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: lets use removeAllCacheEntries? so no need to mention tokens.
please also add a comment here, and some other places that require explanataion

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+      Configuration conf = encryptionConfigurationEntry.getValue();
+
+      Path file = new Path(root, encryptionConfiguration.toString() + ".parquet.encrypted");
+      LOG.info("\nWrite " + file.toString());
+
+      FileEncryptionProperties fileEncryptionProperties = null;
+      if (null == conf) {
+        conf = new Configuration();
+      } else {
+
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: rm empty line

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+      LOG.info("\n\n");
+      LOG.info("==> Decryption configuration {}\n", decryptionConfiguration);
+
+      File folder = new File(root.toString());
+      File[] listOfFiles = folder.listFiles();
+
+      for (int fileNum = 0; fileNum < listOfFiles.length; fileNum++) {
+        KeyToolkit.removeCacheEntriesForToken(KeyToolkit.DEFAULT_ACCESS_TOKEN);
+        Path file = new Path(listOfFiles[fileNum].getAbsolutePath());
+        if (!file.getName().endsWith("parquet.encrypted")) { // Skip non encrypted files
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: don't we want to test reading unencrypted files?

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    "kc1: " + encoder.encodeToString(COLUMN_ENCRYPTION_KEY1) +
+    ", kc2: " +  encoder.encodeToString(COLUMN_ENCRYPTION_KEY2) +
+    ", kf: "  + encoder.encodeToString(FOOTER_ENCRYPTION_KEY);
+  private static final String COLUMN_KEY_MAPPING = "kc1: double_field; kc2: float_field";
+  private static final boolean plaintextFilesAllowed = true;
+
+  final WriteSupport<Group> writeSupport = new GroupWriteSupport();
+
+  @Test
+  public void testWriteReadEncryptedParquetFiles() throws IOException {
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:190]: not only encrypted

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+    Path rootPath = new Path(temporaryFolder.getRoot().getPath());
+    LOG.info("======== testWriteReadEncryptedParquetFiles {} ========", rootPath.toString());
+    // This map will hold various encryption configurations.
+    Map<EncryptionConfiguration, Configuration> encryptionPropertiesMap = getHadoopConfigurationForEncryption();
+    testWriteEncryptedParquetFiles(rootPath, encryptionPropertiesMap);
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:197]: not only encrypted

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+    // This map will hold various decryption configurations.
+    Map<DecryptionConfiguration, Configuration> decryptionPropertiesMap = getHadoopConfigurationForDecryption();
+    testReadEncryptedParquetFiles(rootPath, decryptionPropertiesMap);
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:200]: not only encrypted

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+ *  - ENCRYPT_COLUMNS_AND_FOOTER:   Encrypt two columns and the footer, with different
+ *                                  keys.
+ *  - ENCRYPT_COLUMNS_PLAINTEXT_FOOTER:   Encrypt two columns, with different keys.
+ *                                  Do not encrypt footer (to enable legacy readers)
+ *                                  - plaintext footer mode.
+ *  - ENCRYPT_COLUMNS_AND_FOOTER_CTR:   Encrypt two columns and the footer, with different
+ *                                  keys. Use the alternative (AES_GCM_CTR_V1) algorithm.
+ *  - WRITE_PLAINTEXT:   Do not encrypt anything
+ *
+ *
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:92]: we also need to add a check for writing key material in separate files: eg a loop on the test, with two iterations: a) default b) "encryption.key.material.internal.storage=false".

Plus two other loops: on "encryption.double.wrapping" [a) default b) false] and on ""encryption.wrap.locally" [a) default b) true]

Comment (ggershinsky) [reply parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:92]: and add a key rotation test..

Comment (andersonm-1) [reply parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:92]: Changed the test to parameterized test. 
Key rotation test will be added later.

File: parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java (hunk 6 above)
Comment (andersonm-1) [reply parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: the method is not public

Comment (ggershinsky) [reply parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java]: added a public removeCacheEntriesForAllTokens

**Pull request merged**